packages/react-devtools-inline/__tests__/__e2e__/components.test.js

```diff
@@ -85,21 +85,41 @@ test.describe('Components', () => {
           createTestNameSelector('InspectedElementPropsTree'),
           createTestNameSelector(selectorValue),
         ])[0];
-        const source = findAllNodes(container, [
-          createTestNameSelector('InspectedElementView-Source'),
-        ])[0];
         const value = isEditable.value
           ? valueElement.value
           : valueElement.innerText;
 
-        return [name, value, source.innerText];
+        return [name, value];
       },
       {name: isEditableName, value: isEditableValue}
     );
 
     expect(propName).toBe('label');
     expect(propValue).toBe('"one"');
-    expect(sourceText).toMatch(/e2e-app[a-zA-Z]*\.js/);
+  });
+
+  test('Should allow inspecting source of the element', async () => {
+    // Source inspection is available only in modern renderer.
+    runOnlyForReactRange('>=16.8');
+
+    // Select the first list item in DevTools.
+    await devToolsUtils.selectElement(page, 'ListItem', 'List\nApp');
+
+    // Then read the inspected values.
+    const sourceText = await page.evaluate(() => {
+      const {createTestNameSelector, findAllNodes} = window.REACT_DOM_DEVTOOLS;
```

Comment:
Since this is an E2E test, is there a way we can do this without reaching into the internals?

Reply:
Yes, there is definitely a way to just use query selectors, but this is probably required for more complex cases with inputs

Internals are injected in
https://github.com/facebook/react/blob/850fac4915864a487e7cb9ecae8a75dbac144174/packages/react-devtools-shell/src/e2e/devtools.js#L60-L61

This was added in https://github.com/facebook/react/pull/22978, I think https://github.com/facebook/react/pull/22978#issuecomment-996192672 is one of the reasons.